### PR TITLE
[Merged by Bors] - Added OpaConfig struct

### PR DESCRIPTION
**Pull request by maltesander**
## Description

closes https://github.com/stackabletech/operator-rs/issues/356

## Review Checklist
- [x] Code contains useful comments
- [ ] (Integration-)Test cases added (or not applicable)
- [x] Documentation added (or not applicable)
- [x] Changelog updated (or not applicable)
- [x] Cargo.toml only contains references to git tags (not specific commits or branches)

Once the review is done, comment `bors r+` (or `bors merge`) to merge. [Further information](https://bors.tech/documentation/getting-started/#reviewing-pull-requests)

### CODE REVIEW

File: src/opa.rs [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+    ///
+    /// This may be used if the OPA base url is contained in an ENV variable.
+    ///
+    /// # Example
+    ///
+    /// v1/data/<package>
+    ///
+    /// # Arguments
+    /// * `resource`     - The cluster resource.
+    pub fn package_url<T>(&self, resource: &T) -> String
```
Comment (razvan) [inline src/opa.rs:75]: Personally I'm not a big fan of this: having a (potentially) complex resource as a dependency (parameter) just to read one field from it is usually an anti pattern. Maybe we can introduce a `ClusterName` type or something like that.

Comment (maltesander) [reply src/opa.rs]: Its just something of type `ResourceExt` for the name() and namespace() methods. It defines the behavior as in the doc that we default to the resource name. That would not be true if we have a parameter `cluster_name`.

**Comment by fhennig** [comment]
looks good to me; tests work, docs are nice etc.

But I'll let razvan get back to this

**Comment by fhennig** [comment]
Also we need to wait for github actions to work again :smile:

File: src/opa.rs [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+//! ```
+use crate::client::Client;
+use crate::error;
+use crate::error::OperatorResult;
+use k8s_openapi::api::core::v1::ConfigMap;
+use kube::ResourceExt;
+use schemars::{self, JsonSchema};
+use serde::{Deserialize, Serialize};
+
+const OPA_API: &str = "v1/data";
```
Comment (soenkeliebau) [inline src/opa.rs:56]: I apologize in advance for this comment, I know that I said in the call we are fine to hard code this, but if we do it here I believe we create an artificial requirement for all our authorizers to always use the same version of the data api.
I am perfectly fine to have this hard coded here as a default value (maybe with a different name like 'V1_DATA_API" or something similar that allows for easier addition of other apis later) the individual authorizers have to be able to override this. And by that I do not mean that the user should be able to configure this, but that the authorizer code should be able to do so, still hardcoded, just somewhere else :)

File: src/opa.rs [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+    ///
+    /// This may be used if the OPA base url is contained in an ENV variable.
+    ///
+    /// # Example
+    ///
+    /// v1/data/<package>
+    ///
+    /// # Arguments
+    /// * `resource`     - The cluster resource.
+    pub fn package_url<T>(&self, resource: &T) -> String
```
Comment (soenkeliebau) [inline src/opa.rs:77]: nit: The name of this function does not refer to the api (data) or version of the api (v1) that is called at all. Not a severe issue, but if at some point in time we also want to support the query api or v2 of the data api we might end up with inconsistent naming. 
Comment goes for all other fns as well I guess..

File: src/opa.rs (hunk 2 above)
Comment (maltesander) [reply src/opa.rs]: introduced `OpaApiVersion` enum to differentiate the api version and respective pathes

File: src/opa.rs (hunk 3 above)
Comment (maltesander) [reply src/opa.rs]: renamed to document_url (which reflects the data api) and added the version enum.

File: src/opa.rs [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+    /// This may be used if the OPA base url is contained in an ENV variable.
+    ///
+    /// # Example
+    ///
+    /// * `v1/data/<package>`
+    /// * `v1/data/<package>/<rule>`
+    ///
+    /// # Arguments
+    /// * `resource`     - The cluster resource.
+    /// * `rule`         - The rule name. Defaults to `allow`.
```
Comment (soenkeliebau) [inline src/opa.rs:89]: This is not correct any more I think?

File: src/opa.rs [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+#[derive(Clone, Debug, Default, Deserialize, JsonSchema, PartialEq, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct OpaConfig {
+    pub config_map_name: String,
+    pub package: Option<String>,
+}
+
+impl OpaConfig {
+    /// Returns the OPA data API url. If [`OpaConfig`] has no `package` set,
+    /// will default to the cluster `resource` name.
```
Comment (soenkeliebau) [inline src/opa.rs:78]: Should we also mention what happens if no rule is set?

**Review by soenkeliebau — CHANGES_REQUESTED** [review]

**Review by soenkeliebau — APPROVED** [review]

**Comment by maltesander** [comment]
bors merge

**Pull request closed without merging**